kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java

```diff
@@ -85,10 +102,62 @@ public void forwardRequest(ApiMessage message) {
             throw new AssertionError("Attempt to use forwardRequest with a non-request: " + name);
         }
 
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("{}: Forwarding request: {}", channelDescriptor(), decodedFrame);
+        }
         // TODO check we've not forwarded it already
         channelContext.write(decodedFrame, promise);
     }
 
+    @Override
+    public <T extends ApiMessage> ProxyFuture<T> sendRequest(short apiVersion, ApiMessage message) {
+        checkNotClosed();
+        short key = message.apiKey();
+        var apiKey = ApiKeys.forId(key);
+        short headerVersion = apiKey.requestHeaderVersion(apiVersion);
+        var header = new RequestHeaderData()
+                .setCorrelationId(0)
+                .setRequestApiKey(key)
+                .setRequestApiVersion(apiVersion);
+        if (headerVersion > 1) {
+            header.setClientId(filter.getClass().getSimpleName() + "@" + System.identityHashCode(filter));
+        }
+        boolean hasResponse = apiKey != ApiKeys.PRODUCE
+                || ((ProduceRequestData) message).acks() != 0;
+        var filterPromise = new ProxyPromiseImpl<T>();
+        var frame = DecodedRequestFrame.internalRequest(
+                apiVersion, 0, hasResponse,
+                filter, filterPromise, header, message);
+
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("{}: Sending request: {}", channelDescriptor(), frame);
+        }
+        ChannelPromise writePromise = outboundCtx.voidPromise();
+        // if (outboundCtx.channel().isWritable()) {
+        // outboundCtx.write(frame, writePromise);
+        // }
+        // else {
+        outboundCtx.writeAndFlush(frame, writePromise);
+        // }
+
+        if (!hasResponse) {
+            writePromise.addListener(f -> {
+                // Complete the filter promise for an ack-less Produce
+                // based on the success of the channel write
+                if (f.isSuccess()) {
+                    filterPromise.complete(null);
+                }
+                else {
+                    filterPromise.fail(f.cause());
+                }
+            });
+        }
+        // Otherwise, the filter promise will be completed when handling the response
+
+        // TODO we probably need a timeout mechanism too
```

Comment:
Need to fix this, but probably in a later PR.

Reply:
Could you log a follow-up issue please if it's not going to happen in the scope of #56.